File: hackerrank/brainf-k-interpreter-fp.rb
#!/usr/bin/ruby

class BF
	def initialize(s)
		@inp=s
		@buffer=[0]*99999
		@ptr=0
		@inpptr=0
		@inst=0
	end
	def execute(s)
		i=0
		while i<s.size
			if s[i]=='>'
				@ptr+=1;@inst+=1
			elsif s[i]=='<'
				@ptr-=1;@inst+=1
			elsif s[i]=='+'
				@buffer[@ptr]=(@buffer[@ptr]+1)%256;@inst+=1
			elsif s[i]=='-'
				@buffer[@ptr]=(@buffer[@ptr]-1)%256;@inst+=1
			elsif s[i]=='.'
				@inst+=1
				if @inst>100000
					raise
				end
				print(@buffer[@ptr].chr)
			elsif s[i]==','
				@inst+=1
				if @inst>100000
					raise
				end
				@buffer[@ptr]=@inp[@inpptr].ord;@inpptr+=1
			elsif s[i]=='['
				@inst+=1
				if @buffer[@ptr]!=0
					execute(s[i+1..-1])
				end
				marker=1
				while marker>0
					if s[i+1]=='['
						marker+=1
					elsif s[i+1]==']'
						marker-=1
					end
					i+=1
				end
			elsif s[i]==']'
				@inst+=1
				if @buffer[@ptr]!=0
					@inst+=1
					i=-1
				else
					return
				end
			end
			i+=1
			if @inst>100000
				raise
			end
		end
	end
end

n,m=gets.split.map(&:to_i)
bf=BF.new(gets.chomp.chop)
begin
	bf.execute($<.to_a*'')
rescue
	puts
	puts 'PROCESS TIME OUT. KILLED!!!'
end

File: hackerrank/brainf-k-interpreter-fp.py
#!/usr/bin/python
from __future__ import print_function
import sys
if sys.version_info[0]>=3: raw_input=input

class BF:
	buffer=[0]*99999
	ptr=0
	inpptr=0
	inst=0
	def __init__(self,s):
		self.inp=s
	def execute(self,s):
		i=0
		while i<len(s):
			if s[i]=='>': self.ptr+=1;self.inst+=1
			elif s[i]=='<': self.ptr-=1;self.inst+=1
			elif s[i]=='+': self.buffer[self.ptr]+=1;self.inst+=1
			elif s[i]=='-': self.buffer[self.ptr]-=1;self.inst+=1
			elif s[i]=='.':
				self.inst+=1
				if self.inst>100000:
					raise
				print(chr(self.buffer[self.ptr]),end='')
			elif s[i]==',':
				self.inst+=1
				if self.inst>100000:
					raise
				self.buffer[self.ptr]=ord(self.inp[self.inpptr]);self.inpptr+=1
			elif s[i]=='[':
				self.inst+=1
				if self.buffer[self.ptr]: self.execute(s[i+1:])
				marker=1
				while marker:
					if s[i+1]=='[': marker+=1
					if s[i+1]==']': marker-=1
					i+=1
			elif s[i]==']':
				self.inst+=1
				if self.buffer[self.ptr]:
					self.inst+=1
					i=-1
				else: return
			i+=1
			if self.inst>100000:
				raise

n,m=map(int,raw_input().split())
bf=BF(raw_input().rstrip()[:-1])
prog=''
for _ in range(m):prog+=raw_input()
try:
	bf.execute(prog)
except:
	print()
	print('PROCESS TIME OUT. KILLED!!!')